### DISCUSSION TITLE
Extended State for hisma

### DISCUSSION THREAD

**Discussion by phamnhuvu-dev** [tamas-p/hisma#16] [Q&A] [answered]
As I know, in the document UML State Machine has its Extended State to store the data which the application needs to show on UI for the user.
https://en.wikipedia.org/wiki/UML_state_machine#Extended_states

Do you have any plans for it?
Could I use `bloc` library as an Extended State now?

**Comment by tamas-p** [comment, marked as answer]
First of all thank you for your question. This is one of the areas where I did the minimum effort implementation that satisfies my initial needs for my apps but I am eager to have conversation about how to do it better.

Let's cover first what is there already. Hisma state machines have a dynamic type attribute called `data`. It can be set in the constructor and later by assigning new value to it anytime - as it is not final. Since it is dynamic you can put into this variable whatever complex data structure you want. I just realized that Hisma README.md does not describe machine.data, I will fix this...

[fb_auth_hisma_example](https://github.com/tamas-p/hisma/tree/master/examples/fb_auth_hisma_example) is an example Flutter app that demonstrates how an existing Hisma state machine can be used as the engine for a Flutter app. The machine is [fb_auth_hisma](https://github.com/tamas-p/hisma/tree/master/packages/fb_auth_hisma) that defines a hierarchical state machine for Firebase authentication. One simple example of machine.data usage is in [failed_signin_dialog.dart](https://github.com/tamas-p/hisma/blob/master/examples/fb_auth_hisma_example/lib/app/components/signed_out/components/login/layers/ui/failed_signin_dialog.dart)

When login fails an error message (that is coming from the machine) is shown in a dialog. This happens as follows:
- [login_machine.dart](https://github.com/tamas-p/hisma/blob/master/packages/fb_auth_hisma/lib/src/components/signed_out/components/login/layers/machines/login_machine.dart): In _createEmailSignInAction() login fails.
- Exception's code is used as the arg for machine.fire(ELiM.fail, arg: e.code) to take machine to SLiM.failedSignIn state.
- In onEntry action of the SLiM.failedSignIn state we set machine.data to the arg (that is now the exception's code) we received through the fire operation above.
- hisma_flutter router shows the failedSignInDialog when machine state changed to SLiM.failedSignIn
- failedSignInDialog uses machine.data as the message to be shown

One could use radically different approach as well where data storage is completely independent from the state machine, like a data binding library e.g. Riverpod. However, in that case you would need to have your provider known by both the machine and the ui (so to your question I think you could use bloc - I just only have experience with with Riverpod when prototyped apps with Hisma). 

In some cases it it could be perfectly fine with all the benefits the data binding library provides, but I partially introduced machine.data to have a simple built-in approach to transferring data from the machine layer to the UI layer where machine is independent from the data binding choices of the UI layer. Another use is what mentioned in the Wikipedia article you referenced: having data attached to the machine that is required for its operation. This other use-case is that I extensively use e.g. in the unit tests for features like minInterval, guards and onEntry, onExit actions (e.g.  [transition_interval_test.dart](https://github.com/tamas-p/hisma/blob/master/packages/hisma/test/transition_interval_test.dart) ).

All in all, I think it makes sense to keep and improve the approach when data is part of the state machine itself and any comments, ideas are more than welcomed.

**Reply by tamas-p** [reply]
also see #14 (nothing more there yet, but that's the issue I entered to track this in January)

**Reply by phamnhuvu-dev** [reply]
Thank you for your answer, I will read more about your project, your visualization tool with plantuml is awesome.
I am developing Model Driven Development with plantuml for my team to make the project able to visualize => and make complexity manageable.
So I need to check both the fundamentals, concepts, and principles of HSM and your implementation to make sure that it is good enough to run on the production.

Thank you for your contribution.

**Reply by tamas-p** [reply]
*"your visualization tool with plantuml is awesome"*
Thank you!

*"I am developing Model Driven Development with plantuml"*
That is interesting. In case you do not already have your dart abstraction to plantuml consider using [pumli](https://github.com/tamas-p/pumli) that is a package created for visma.

"and make complexity manageable."
That's the key here, I agree. Is your tooling is public?

**Reply by phamnhuvu-dev** [reply]
It is not public, because I am focusing on productivity in my team first because it depends on the tech stacks of the company.
For example:
I draw a class in plantuml and I generate it to the code, but in the code, we use built_value to build entity classes, so they are different

the techniques and mindset are popular, and a lot of tools support generating code from structure UML diagrams.

But for Behavior UML Diagrams, only State Machine diagram is able to do that now when I know this project https://stately.ai/

Don't try to port XState from Javascript to Dart, because Javascript is a dynamic language, it is very difficult to port to a more strict language like Dart

So my idea is, In the system, there are a lot of elements, they have relationships and behaviors with each other, and we have to make them able to visualize to make complexity manageable.
The developers have to spend time designing before implementing, not just receiving requirements and then planning and coding.

But to reduce the time of the development cycle, this tool has to support generating code.

**Comment by phamnhuvu-dev** [comment]
Just sharing some interesting things.
https://github.com/google/hierarchical-state-machine.dart
https://developers.redhat.com/articles/2021/11/23/how-design-state-machines-microservices#
